sphinx/ext/autodoc/__init__.py

```diff
@@ -1449,7 +1449,7 @@ def add_directive_header(self, sig: str) -> None:
         super().add_directive_header(sig)
         sourcename = self.get_sourcename()
         if inspect.isabstractmethod(self.object):
-            self.add_line('   :abstractmethod:', sourcename)
+            self.add_line('   :abstract:', sourcename)
```

Comment:
Since properties are not methods, having the following doesn’t make sense:

```restructuredtext
.. attribute:: thing
   :abstractmethod:
```

The correct way to create an abstract property is of course still

```py
@property
@abstractmethod
def thing(self):
   ...
```

but I see it this way:

1. `def thing(self):` without decorators is a method
2. adding `@abstractmethod` makes a method into an abstract method
3. adding `@property` makes a method into a property (and therefore an abstract method into an abstract property)

Of course I’d also have called `.. method::`’s option `:abstract:` instead of `:abstractmethod:`, but it’s too late for that.

Reply:
-1: It's inconsistent. If we choose the way to options describes by its character, we should re-design all of options like `:staticmethod:`, `:abstractmethod:`, `:classmethod:` and so on. This is one of reason why we chose to define properties by "how to defined". I believe this PR is a changing of policy. We have to design whole of designs.

BTW, we use `@abstractmethod` to define abstract objects in python. And we use `:abstractmethod:` for methods. On the other hand, we should use `:abstract:` only for properties. It is difficult to learn how to mark up.

Reply:
Can we use only `abstract` option without `property`?

Reply:
> Can we use only `abstract` option without `property`?

I think so:

```py
class AbstractFruit(ABC):
    color: Color

class Orange(AbstractFruit):
    color = Color.orange

class Apple(AbstractFruit):
	def __init__(self, ripe: bool):
        self.ripe = ripe
    @property
    def color(self):
        return Color.red if self.ripe else Color.green
```

- `AbstractFruit` has an abstract attribute
- `Orange` overrides it with a concrete attribute
- `Apple` overrides it with a concrete property

Reply:
> If we choose the way to options describes by its character, we should re-design all of options like :staticmethod:, :abstractmethod:, :classmethod: and so on.

`:abstract:` could just become an alias for `:abstractmethod:` (make both work for methods). Alternatively we could call the version for `.. attribute::` just `:abstractattribute:` or so.

Reply:
I'll call `AbstractFruit.color` as "an instance attribute of the abstract class" if I need to mention to it. But I understand it is an abstract attribute. So it will make a sense.

>:abstract: could just become an alias for :abstractmethod: (make both work for methods). Or we keep :abstractmethod: and call the version for .. attribute:: :abstractattribute: or similar.

If we choose former one, `:class:` and `:static:` are also needed, right?  But we already use `:class:` for styling. So they're conflicted...

Reply:
We could just allow `:abstract:` as well as `:abstractmethod:` for methods. That way nobody is confused that one or the other doesn’t work.

Reply:
I prefer "Zen of Python". So there should be one obvious way to do it. Of course, we need to provide an old options to keep compatibility. But the ideal option should be one. So we should choose a new option from these candidates. -1 for adding both.

If we choose `:abstract:` for attributes, other options will be also changed to describe its behavior. So methods and attributes will be marked up as following:
```
.. py:method:: hello()
   :abstract:
   :async:
   :static:
   :classmethod:

   Note: ``:class:`` is not able to use to describe classmethod
   because we have already have ``:class:`` option for commonly used options in reST directives.

.. py:attribute:: name
   :abstract:
   :async:
   :property:
```

If we choose `:abstractattribute:`, they are marked up as following:
```
.. py:method:: hello()
   :abstractmethod:
   :async:
   :staticmethod:
   :classmethod:

.. py:attribute:: name
   :abstractattribute:
   :async:
   :property:
```
Is my understanding correct?

I feel former one is not consistent because of `:classmethod:`. On the other hand, latter one is a bit non natural to me because `:abstractattribute:` is not a keyword of python.

Reply:
Just idea, it might be better to add a new directive `py:property`. It helps to separate options from `py:attribute`:
```
.. py:attribute:: name
   :type:
   :value:

.. py:property:: name
   :abstractmethod:
   :async:
```

Reply:
IMO, as a document writer, it's easy to use the terms in Python itself. Otherwise, if sphinx require names like `abstract`,` static`, `class`, etc., document writers will need to remember the names for writing documentation in Sphinx separately from the Python terminology. For this reason, `:abstractmethod:` is better than the common option `:abstract:`.

By the way, your `AbstractFruit` is instanciable...
```
>>> class AbstractFruit(ABC):
...     color: Color
...
>>> af = AbstractFruit()
>>> af
<__main__.AbstractFruit object at 0x0000014C8D051588>
```
If we want to create such class, it is:
```
>>> class AbstractFruit(ABC):
...     @property
...     @abstractmethod
...     def color(self) -> Color:
...         pass
...
>>> af = AbstractFruit()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: Can't instantiate abstract class AbstractFruit with abstract methods color
```
I think this `color` attribute is easy to understand in the documentation as follows:
```
.. py:method:: color
  :abstractmethod:
  :property:

  describing string

  :return: color value
  :rtype: Color
```

To be honest, it would be nice to be able to write short as:
```
.. py:method:: color
  :abstractmethod:
  :property:
  :type: Color
```

Reply:
Let’s continue discussing below, OK? @shimizukawa I understand what you mean about mirroring the code. Since `.. py:method:: […] :property:` doesn’t make sense though, I assume you’d vote for the following?

```restructuredtext
.. py:attribute:: color
  :abstractmethod:
  :property:
```

Reply:
> Let’s continue discussing below, OK? 

sure :)